In [22]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from cmdstanpy import CmdStanModel

import arviz as az

In [23]:
# Get data
data = pd.read_csv('data/recognition.csv', header=0, sep=',')
data.head()

# Prepare data in Stan friendly format
data_dict = {
    'N': data.shape[0],
    'old': data.old.values,
    'response': data.response.values
}

In [30]:
# Compile 1HT model
model1 = CmdStanModel(stan_file="hw4_2_1.stan")

In [25]:
# Sample 1HT model
fit1 = model1.sample(data=data_dict, chains=4, iter_sampling=2500, iter_warmup=1000)

# Display sampling diagnostics
print(fit1.diagnose())

14:55:08 - cmdstanpy - INFO - CmdStan start processing
chain 1:   0%|          | 0/3500 [00:00<?, ?it/s, (Warmup)]



chain 1:  29%|██▊       | 1000/3500 [00:00<00:00, 15552.49it/s, (Sampling)]

chain 1:  60%|██████    | 2100/3500 [00:00<00:00, 20974.82it/s, (Sampling)]


chain 2: 100%|██████████| 3500/3500 [00:00<00:00, 19323.55it/s, (Sampling completed)]

chain 3: 100%|██████████| 3500/3500 [00:00<00:00, 19300.89it/s, (Sampling completed)]


chain 4: 100%|██████████| 3500/3500 [00:00<00:00, 19267.25it/s, (Sampling completed)]


14:55:09 - cmdstanpy - INFO - CmdStan done processing.



Checking sampler transitions treedepth.
Treedepth satisfactory for all transitions.

Checking sampler transitions for divergences.
No divergent transitions found.

Checking E-BFMI - sampler transitions HMC potential energy.
E-BFMI satisfactory.

Rank-normalized split effective sample size satisfactory for all parameters.

Rank-normalized split R-hat values satisfactory for all parameters.

Processing complete, no problems detected.



In [26]:
# Compile 2HT model
model2 = CmdStanModel(stan_file="hw4_2_2.stan")

In [27]:
# Sample 2HT model
fit2 = model2.sample(data=data_dict, chains=4, iter_sampling=2500, iter_warmup=1000)

# Display sampling diagnostics
print(fit2.diagnose())

14:55:09 - cmdstanpy - INFO - CmdStan start processing
chain 1:   0%|          | 0/3500 [00:00<?, ?it/s, (Warmup)]


chain 1:  29%|██▊       | 1000/3500 [00:00<00:00, 9972.57it/s, (Warmup)]


chain 1:  29%|██▊       | 1000/3500 [00:00<00:00, 9972.57it/s, (Sampling)]


chain 1:  80%|████████  | 2800/3500 [00:00<00:00, 14390.95it/s, (Sampling)]


chain 2: 100%|██████████| 3500/3500 [00:00<00:00, 13891.34it/s, (Sampling completed)]

chain 3: 100%|██████████| 3500/3500 [00:00<00:00, 13889.63it/s, (Sampling completed)]


chain 4: 100%|██████████| 3500/3500 [00:00<00:00, 13854.46it/s, (Sampling completed)]


14:55:09 - cmdstanpy - INFO - CmdStan done processing.



Checking sampler transitions treedepth.
Treedepth satisfactory for all transitions.

Checking sampler transitions for divergences.
No divergent transitions found.

Checking E-BFMI - sampler transitions HMC potential energy.
E-BFMI satisfactory.

Rank-normalized split effective sample size satisfactory for all parameters.

Rank-normalized split R-hat values satisfactory for all parameters.

Processing complete, no problems detected.



In [28]:
# Display 1HT sampling data
summary = fit1.summary()
display(summary)

draws_df = fit1.draws_pd()
display(draws_df.head())

,Mean,MCSE,StdDev,MAD,5%,50%,95%,ESS_bulk,ESS_tail,ESS_bulk/s,R_hat
lp__,-23.296800,0.014626,1.006830,0.716300,-25.357300,-22.986700,-22.341700,4873.03,6394.11,12756.6,1.00010
d,0.674935,0.001458,0.123169,0.125456,0.455127,0.686738,0.854815,7135.40,6869.21,18679.1,1.00058
g,0.347189,0.001227,0.106554,0.110974,0.179783,0.344721,0.528232,7468.88,6422.61,19552.0,1.00029


,chain__,iter__,draw__,lp__,accept_stat__,stepsize__,treedepth__,n_leapfrog__,divergent__,energy__,d,g
0,1.0,1.0,1.0,-22.946391,0.899685,0.793829,1.0,1.0,0.0,22.946911,0.565536,0.287604
1,1.0,2.0,2.0,-22.922439,0.996372,0.793829,2.0,3.0,0.0,23.350481,0.797447,0.281455
2,1.0,3.0,3.0,-24.832699,0.832391,0.793829,2.0,3.0,0.0,25.249214,0.896010,0.312322
3,1.0,4.0,4.0,-25.363586,0.971885,0.793829,2.0,3.0,0.0,25.758608,0.902210,0.395853
4,1.0,5.0,5.0,-22.433240,0.994590,0.793829,2.0,7.0,0.0,24.905553,0.619200,0.327990


In [29]:
# Display 2HT sampling data
summary = fit2.summary()
display(summary)

draws_df = fit2.draws_pd()
display(draws_df.head())

,Mean,MCSE,StdDev,MAD,5%,50%,95%,ESS_bulk,ESS_tail,ESS_bulk/s,R_hat
lp__,-20.651800,0.014914,1.002480,0.728511,-22.675400,-20.349500,-19.692400,4633.83,6138.71,8440.49,1.00038
d,0.672816,0.001257,0.113265,0.114490,0.469044,0.683004,0.841563,8128.86,6403.74,14806.70,1.00034
g,0.547151,0.001461,0.136613,0.144023,0.314714,0.550801,0.765531,8668.72,6214.03,15790.00,1.00054


,chain__,iter__,draw__,lp__,accept_stat__,stepsize__,treedepth__,n_leapfrog__,divergent__,energy__,d,g
0,1.0,1.0,1.0,-20.966607,0.984127,0.881096,2.0,3.0,0.0,22.790909,0.636458,0.762267
1,1.0,2.0,2.0,-20.480631,0.736313,0.881096,1.0,3.0,0.0,24.185378,0.795515,0.456614
2,1.0,3.0,3.0,-20.154009,1.000000,0.881096,2.0,3.0,0.0,20.775908,0.671108,0.691766
3,1.0,4.0,4.0,-20.051719,0.975182,0.881096,2.0,7.0,0.0,20.617419,0.761549,0.617702
4,1.0,5.0,5.0,-20.660348,0.940266,0.881096,2.0,7.0,0.0,20.844526,0.616586,0.362323


The One-High-Threshold model estimates `d` with a mean of `0.675` and `g` with a mean of `0.347`.
The Two-High-Threshold model estimates `d` with a mean of `0.673` and `g` with a mean of `0.547`.

The same `d` value implies that both models have the same confidence in selecting
the correct answer. But, the 1HT model having a lower `g` value implies that the model, when unsure of the correct answer, will be more likely to respond that the word is new.
The `g` value for the 2HT model implies that the model has more of a 50/50 chance of
picking either reponse when unsure.

This most likely occurs because the 1HT model assumes that when looking at a new word, 
the person responding will be unsure of their guess. But, that isn't necessarily the case,
so if the data shows that new words were correctly labeled as new more often than old,
then it would push the value of `g` down to account for that.

But the 2HT model does not treat all responses to new words as guesses, so the actual
guessing part is more of a 50/50.